lib/fluent/plugin/out_mongo.rb

```diff
@@ -227,35 +184,26 @@ def format_collection_name(collection_name)
       formatted
     end
 
-    def get_or_create_collection(collection_name)
-      collection_name = format_collection_name(collection_name)
-      return @clients[collection_name] if @clients[collection_name]
-
-      @db ||= get_connection
-      if @db.collection_names.include?(collection_name)
-        collection = @db.collection(collection_name)
-        unless @disable_collection_check
-          capped = collection.capped?
-          unless @collection_options[:capped] == capped # TODO: Verify capped configuration
-            new_mode = format_collection_mode(@collection_options[:capped])
-            old_mode = format_collection_mode(capped)
-            raise ConfigError, "New configuration is different from existing collection: new = #{new_mode}, old = #{old_mode}"
+    def operate(collection, records)
+      begin
+        if @replace_dot_in_key_with
+          records.map! do |r|
+            replace_key_of_hash(r, ".", @replace_dot_in_key_with)
+          end
+        end
+        if @replace_dollar_in_key_with
+          records.map! do |r|
+            replace_key_of_hash(r, /^\$/, @replace_dollar_in_key_with)
           end
         end
-      else
-        collection = @db.create_collection(collection_name, @collection_options)
-      end
-
-      @clients[collection_name] = collection
-    end
-
-    def format_collection_mode(mode)
-      mode ? 'capped' : 'normal'
-    end
 
-    def get_connection
-      db = Mongo::MongoClient.new(@host, @port, @connection_options).db(@database)
-      authenticate(db)
+        @client[collection, @collection_options].insert_many(records)
```

Comment:
For invalid encoding records, I think we cannot obtain additional information from Exception.
Because invalid encoding records causes `ArgumentError` exception which raised from BSON, not `Mongo::Error::BulkWriteError` which raised from mongo ruby driver.

```ruby
# coding: utf-8
require 'mongo'
client = Mongo::Client.new([ '127.0.0.1:27017' ], :database => 'test')
collection = client[:people]
doc = [{'a' => 3, 'b' => "c", '$last' => '石動'}, {'a' => 4, 'b' => "d", 'first' => '菖蒲'.encode('EUC-JP').force_encoding('UTF-8')}]
begin
  client[collection].insert_many(doc)
rescue Mongo::Error::BulkWriteError => e
  p e.class
rescue ArgumentError => e
  p e.class
end
```

```log
$ bundle exec ruby insert_invalid_record.rb 
D, [2016-10-04T11:57:16.491145 #9648] DEBUG -- : MONGODB | Adding 127.0.0.1:27017 to the cluster.
D, [2016-10-04T11:57:16.506306 #9648] DEBUG -- : MONGODB | 127.0.0.1:27017 | test.insert | STARTED | {"insert"=>"#<Mongo::Collection:0x007fabb52935c0>", "documents"=>[{"a"=>3, "b"=>"c", "$last"=>"石動", :_id=>BSON::ObjectId('57f31a8cb27e1c25b0453fb7')}, {"a"=>4, "b"=>"d", "first"=>"\xBEԳ\xF7", :_id=>BSON::ObjectId('57f31a8cb27e1c25b0453fb8')}], "ordere...
D, [2016-10-04T11:57:16.506620 #9648] DEBUG -- : MONGODB | 127.0.0.1:27017 | test.insert | FAILED | String ?Գ? is not valid UTF-8. | 0.000149s
ArgumentError
```

For `Mongo::Error::BulkWriteError`, I think we can obtain it.
I'll more investigate it.